### PR TITLE
feat: implement MaxCollateralShare

### PR DESCRIPTION
**Pull request by toteki**
## Description

closes: #1093 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by RafilxTenfen — APPROVED** [review]

File: x/leverage/keeper/collateral.go [hunk 1]
```diff
@@ -89,14 +88,14 @@ func (k Keeper) setCollateralAmount(ctx sdk.Context, borrowerAddr sdk.AccAddress
 
 // GetTotalCollateral returns an sdk.Coin representing how much of a given uToken
 // the x/leverage module account currently holds as collateral. Non-uTokens return zero.
-func (k Keeper) GetTotalCollateral(ctx sdk.Context, denom string) sdkmath.Int {
+func (k Keeper) GetTotalCollateral(ctx sdk.Context, denom string) sdk.Coin {
```
Comment (robert-zaremba) [inline x/leverage/keeper/collateral.go:91]: How about renaming this to `GetCollateral` and `GetAllTotalCollateral` to `GetTotalCollateral`?

File: x/leverage/keeper/collateral.go [hunk 2]
```diff
@@ -124,3 +123,15 @@ func (k Keeper) CalculateCollateralValue(ctx sdk.Context, collateral sdk.Coins)
...
+func (k Keeper) GetAllTotalCollateral(ctx sdk.Context) sdk.Coins {
+	total := sdk.NewCoins()
+
+	tokens := k.GetAllRegisteredTokens(ctx)
+	for _, t := range tokens {
+		uDenom := types.ToUTokenDenom(t.BaseDenom)
+		total = total.Add(k.GetTotalCollateral(ctx, uDenom))
+	}
+	return total
+}
```
Comment (robert-zaremba) [inline x/leverage/keeper/collateral.go:137]: I think we should cache it in the state. Something for v4 or later.

File: x/leverage/keeper/keeper.go [hunk 3]
```diff
@@ -274,33 +274,43 @@ func (k Keeper) Repay(ctx sdk.Context, borrowerAddr sdk.AccAddress, payment sdk.
...
+	max, err := k.maxCollateral(ctx, uToken.Denom)
+	if err != nil {
 		return err
 	}
 
-	return k.bankKeeper.SendCoinsFromAccountToModule(ctx, borrowerAddr, types.ModuleName, sdk.NewCoins(coin))
+	total := k.GetTotalCollateral(ctx, uToken.Denom)
+	if total.Add(uToken).Amount.GT(max) {
+		return types.ErrMaxCollateralShare
+	}
```
Comment (robert-zaremba) [inline x/leverage/keeper/keeper.go:290]: We decided to implement this as a `%`

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
We decided to implement MaxCollateralShare as a % value

File: x/leverage/keeper/collateral.go (hunk 2 above)
Comment (robert-zaremba) [reply x/leverage/keeper/collateral.go:137]: https://github.com/umee-network/umee/issues/1336

File: x/leverage/keeper/collateral.go (hunk 1 above)
Comment (robert-zaremba) [reply x/leverage/keeper/collateral.go:91]: After thinking more I think the current name is OK, but lmk what do you think

File: x/leverage/keeper/collateral.go [hunk 4]
```diff
@@ -124,3 +123,15 @@ func (k Keeper) CalculateCollateralValue(ctx sdk.Context, collateral sdk.Coins)
 
 	return limit, nil
 }
+
+// GetAllTotalCollateral returns total collateral across all uToken denominations.
```
Comment (robert-zaremba) [inline x/leverage/keeper/collateral.go:127]:
```suggestion
// GetAllTotalCollateral returns total collateral across all uTokens.
```

Comment (robert-zaremba) [reply x/leverage/keeper/collateral.go:127]: https://www.merriam-webster.com/dictionary/denomination

File: x/leverage/keeper/collateral.go (hunk 1 above)
Comment (toteki) [reply x/leverage/keeper/collateral.go:91]: Current functions that exist:
`GetAllTotalCollateral()`
`GetTotalCollateral(denom)`
`GetBorrowerCollateral(address)`
`GetCollateral(address,denom)`

`GetCollateral` already exists so we'd have to shuffle other things around. That can be saved for a refactor PR if we decide to do it.

File: x/leverage/keeper/keeper.go [hunk 5]
```diff
@@ -274,33 +274,43 @@ func (k Keeper) Repay(ctx sdk.Context, borrowerAddr sdk.AccAddress, payment sdk.
...
 
-	currentCollateral := k.GetCollateralAmount(ctx, borrowerAddr, coin.Denom)
-	if err := k.setCollateralAmount(ctx, borrowerAddr, currentCollateral.Add(coin)); err != nil {
+	max, err := k.maxCollateral(ctx, uToken.Denom)
+	if err != nil {
 		return err
 	}
 
-	return k.bankKeeper.SendCoinsFromAccountToModule(ctx, borrowerAddr, types.ModuleName, sdk.NewCoins(coin))
+	total := k.GetTotalCollateral(ctx, uToken.Denom)
```
Comment (robert-zaremba) [inline x/leverage/keeper/keeper.go:287]: `k.maxCollateral` already computes that, would be good to do it once - maybe we can return it from `k.maxCollateral`?

File: x/leverage/keeper/keeper.go (hunk 3 above)
Comment (robert-zaremba) [reply x/leverage/keeper/keeper.go:290]: OK, I misread it - in `k.maxCollateral` we are using `%` (share value)

File: x/leverage/keeper/keeper.go [hunk 6]
```diff
@@ -274,33 +274,43 @@ func (k Keeper) Repay(ctx sdk.Context, borrowerAddr sdk.AccAddress, payment sdk.
...
+		return err
+	}
+
+	return k.bankKeeper.SendCoinsFromAccountToModule(ctx, borrowerAddr, types.ModuleName, sdk.NewCoins(uToken))
 }
 
 // Decollateralize disables selected uTokens for use as collateral by a single borrower.
-func (k Keeper) Decollateralize(ctx sdk.Context, borrowerAddr sdk.AccAddress, coin sdk.Coin) error {
-	if err := k.validateUToken(coin); err != nil {
+func (k Keeper) Decollateralize(ctx sdk.Context, borrowerAddr sdk.AccAddress, uToken sdk.Coin) error {
```
Comment (robert-zaremba) [inline x/leverage/keeper/keeper.go:301]: :+1: for uToken

File: x/leverage/keeper/suite_test.go [hunk 7]
```diff
@@ -97,6 +97,12 @@ func coin(denom string, amount int64) sdk.Coin {
 	return sdk.NewInt64Coin(denom, amount)
 }
 
+// registerToken adds or updates a token in the token registry and requires no error.
+func (s *IntegrationTestSuite) registerToken(token types.Token) {
+	app, ctx, require := s.app, s.ctx, s.Require()
+	require.NoError(app.LeverageKeeper.SetTokenSettings(ctx, token))
```
Comment (robert-zaremba) [inline x/leverage/keeper/suite_test.go:103]: no need to copy `app` and `ctx` here

**Review by robert-zaremba — COMMENTED** [review]
overall looks good. Let's don't calculate `GetTotalCollateral` twice.

File: x/leverage/keeper/collateral.go (hunk 1 above)
Comment (robert-zaremba) [reply x/leverage/keeper/collateral.go:91]: BTW, renaming is super easy in most of the editors - just few clicks.

**Review by robert-zaremba — APPROVED** [review]
pre-approving. Please look at the comments.

File: x/leverage/keeper/keeper.go (hunk 5 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go:287]: There are a few efficiency upgrades we can do here (and also in the market summary query) but I'm not sure how far we should optimize in this PR.

Another option would be to make a function `k.CollateralShare(denom)` and use it in an error check at the _end_ of `k.Collateralize`, while continuing to use function `maxCollateral` in `grpc_query.go`.

I ended up avoiding that optimization to keep the code smaller, but would prefer it if we have to optimize now.

Comment (toteki) [reply x/leverage/keeper/keeper.go]: Refactored to removed duplicate

**Pull request merged**